chain-indexer/src/domain/ledger_state.rs

```diff
@@ -190,6 +220,21 @@ impl LedgerState {
 
 #[derive(Debug, Error)]
 pub enum Error {
-    #[error("cannot apply transaction")]
-    ApplyTransaction(#[from] indexer_common::domain::ledger::Error),
+    #[error("cannot apply {variant} transaction {hash}")]
+    ApplyTransaction {
```

Comment:
As mentioned above:
- Split into two variants for regular and system tx
- Use tuple structs, i.e. remove the fields; access them in the error string with {0}, ...
- Instead of turning into a string at the call site, make the variant generic `where T: ToString` (or `T: Display` it that does not work); {hash} (or {0}) will then just work.